tools/firmware-utils/src/ptgen.c

```diff
@@ -31,15 +31,62 @@
 #include <ctype.h>
```

Comment:
clang-10:

```make
mkdir -p out
clang-10 -O2 -I/opt/devel/openwrt/openwrt.git/staging_dir/host/include -L/opt/devel/openwrt/openwrt.git/staging_dir/host/lib -Wall -Werror -Wextra -Wno-unused-parameter -include endian.h -o out/ptgen ptgen.c cyg_crc32.c 
ptgen.c:384:6: error: format specifies type 'long long' but the argument has type 'unsigned long' [-Werror,-Wformat]
                                        start * DISK_SECTOR_SIZE, sect * DISK_SECTOR_SIZE,
                                        ^~~~~~~~~~~~~~~~~~~~~~~~
ptgen.c:384:32: error: format specifies type 'long long' but the argument has type 'unsigned long' [-Werror,-Wformat]
                                        start * DISK_SECTOR_SIZE, sect * DISK_SECTOR_SIZE,
                                                                  ^~~~~~~~~~~~~~~~~~~~~~~
ptgen.c:385:6: error: format specifies type 'long long' but the argument has type 'unsigned long' [-Werror,-Wformat]
                                        (sect - start) * DISK_SECTOR_SIZE);
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ptgen.c:386:20: error: format specifies type 'long long' but the argument has type 'unsigned long' [-Werror,-Wformat]
                printf("%lld\n", start * DISK_SECTOR_SIZE);
                        ~~~~     ^~~~~~~~~~~~~~~~~~~~~~~~
                        %lu
ptgen.c:387:20: error: format specifies type 'long long' but the argument has type 'unsigned long' [-Werror,-Wformat]
                printf("%lld\n", (sect - start) * DISK_SECTOR_SIZE);
                        ~~~~     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                        %lu
ptgen.c:374:40: error: comparison of integers of different signs: 'unsigned int' and 'int' [-Werror,-Wsign-compare]
                if (parts[i].type == 0xEF || (i + 1) == active) {
                                              ~~~~~  ^  ~~~~~~
6 errors generated.
Makefile:17: recipe for target 'all' failed
make: *** [all] Error 1
```

gcc-9:

```make
mkdir -p out
gcc-9 -O2 -I/opt/devel/openwrt/openwrt.git/staging_dir/host/include -L/opt/devel/openwrt/openwrt.git/staging_dir/host/lib -Wall -Werror -Wextra -Wno-unused-parameter -include endian.h -o out/ptgen ptgen.c cyg_crc32.c 
ptgen.c: In function ‘init_utf16’:
ptgen.c:253:59: error: suggest parentheses around arithmetic in operand of ‘|’ [-Werror=parentheses]
  253 |    buf[i] = cpu_to_le16((str[n] & 0x1F) << 6 | str[n + 1] & 0x3F);
      |                                                ~~~~~~~~~~~^~~~~~
ptgen.c:42:25: note: in definition of macro ‘cpu_to_le16’
   42 | #define cpu_to_le16(x) (x)
      |                         ^
ptgen.c:256:87: error: suggest parentheses around arithmetic in operand of ‘|’ [-Werror=parentheses]
  256 |    buf[i] = cpu_to_le16((str[n] & 0x0F) << 12 | (str[n + 1] & 0x3F) << 6 | str[n + 2] & 0x3F);
      |                                                                            ~~~~~~~~~~~^~~~~~
ptgen.c:42:25: note: in definition of macro ‘cpu_to_le16’
   42 | #define cpu_to_le16(x) (x)
      |                         ^
ptgen.c: In function ‘gen_gptable’:
ptgen.c:374:40: error: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘int’ [-Werror=sign-compare]
  374 |   if (parts[i].type == 0xEF || (i + 1) == active) {
      |                                        ^~
ptgen.c:376:46: error: taking address of packed member of ‘struct gpte’ may result in an unaligned pointer value [-Werror=address-of-packed-member]
  376 |    init_utf16("EFI System Partition", gpte[i].name, GPT_ENTRY_NAME_SIZE / sizeof(uint16_t));
      |                                       ~~~~~~~^~~~~
ptgen.c:382:44: error: format ‘%lld’ expects argument of type ‘long long int’, but argument 4 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Werror=format=]
  382 |    fprintf(stderr, "Partition %d: start=%lld, end=%lld, size=%lld\n",
      |                                         ~~~^
      |                                            |
      |                                            long long int
      |                                         %ld
ptgen.c:382:54: error: format ‘%lld’ expects argument of type ‘long long int’, but argument 5 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Werror=format=]
  382 |    fprintf(stderr, "Partition %d: start=%lld, end=%lld, size=%lld\n",
      |                                                   ~~~^
      |                                                      |
      |                                                      long long int
      |                                                   %ld
ptgen.c:382:65: error: format ‘%lld’ expects argument of type ‘long long int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Werror=format=]
  382 |    fprintf(stderr, "Partition %d: start=%lld, end=%lld, size=%lld\n",
      |                                                              ~~~^
      |                                                                 |
      |                                                                 long long int
      |                                                              %ld
ptgen.c:386:14: error: format ‘%lld’ expects argument of type ‘long long int’, but argument 2 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Werror=format=]
  386 |   printf("%lld\n", start * DISK_SECTOR_SIZE);
      |           ~~~^
      |              |
      |              long long int
      |           %ld
ptgen.c:387:14: error: format ‘%lld’ expects argument of type ‘long long int’, but argument 2 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Werror=format=]
  387 |   printf("%lld\n", (sect - start) * DISK_SECTOR_SIZE);
      |           ~~~^
      |              |
      |              long long int
      |           %ld
ptgen.c:391:79: error: operand of ?: changes signedness from ‘int’ to ‘long unsigned int’ due to unsignedness of other operand [-Werror=sign-compare]
  391 |  gpte[GPT_ENTRY_MAX - 1].end = cpu_to_le64((kb_align ? round_to_kb(sectors) : sectors) - 1);
      |                                                                               ^~~~~~~
ptgen.c:44:25: note: in definition of macro ‘cpu_to_le64’
   44 | #define cpu_to_le64(x) (x)
      |                         ^
cc1: all warnings being treated as errors
Makefile:16: recipe for target 'all' failed
make: *** [all] Error 1
```

Reply:
Those format issues should be probably handled via PRI*64 macro formatting helpers.